### PR TITLE
[JUJU-1346] Streamline reflection logic

### PR DESCRIPTION
**Pull request by manadart**
The `reflect` package now represents the fact that Sqlair does no struct field inference for fields without a `db` tag. It has been also been refactored for terseness.

A singleton instance of the cache is enforced by the library.

The original query logic, now invalidated by the changes has been removed.

### CODE REVIEW

File: reflect/cache.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// generate
+func generate(value reflect.Value) (Kind, error) {
+	// Dereference the pointer if it is one.
+	value = reflect.Indirect(value)
+
+	// If this is a not a struct, we can not provide
+	// any further reflection information.
+	if value.Kind() != reflect.Struct {
```
Comment (manadart) [inline reflect/cache.go:47]: I will come back around to this when I think about usage of the proposed `sqlair.M` as a new type for `map[string]any`.

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
 module github.com/SimonRichardson/sqlair
```
Comment (juanmanuel-tirado) [inline go.mod]: To be changed.

File: reflect/singleton.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+var (
+	singleCache *cache
+	once        sync.Once
+)
+
+// Cache enforces the singleton pattern,
+// ensuring access to a single instance of cache.
+func Cache() *cache {
+	once.Do(func() {
```
Comment (juanmanuel-tirado) [inline reflect/singleton.go:16]: If we are going to force `Cache` to follow a singleton pattern I would include this code into `reflect/cache.go`.

**Review by juanmanuel-tirado — COMMENTED** [review]
Some minor comments.

File: go.mod (hunk 2 above)
Comment (manadart) [reply go.mod]: Done.

File: reflect/singleton.go (hunk 3 above)
Comment (manadart) [reply reflect/singleton.go:16]: I actually like this separation of cache behaviour and instantiation, like you would do for a factory implementation.

Comment (juanmanuel-tirado) [reply reflect/singleton.go:16]: I understand your point. I'm not a big fan of separating instantiation logic if it only impacts one entity like in this case. When the code base grows you have to look for a very specific struct instantiation far from its definition. Said that, I'm OK with that if we follow that logic in the project.

**Review by juanmanuel-tirado — APPROVED** [review]
LG2M :+1:

**Pull request merged**